Source/WebDriver/Session.cpp

```diff
@@ -3144,4 +3145,183 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
     });
 }
 
+#if ENABLE(WEBDRIVER_BIDI)
+void Session::dispatchEvent(RefPtr<JSON::Object>&& message)
+{
+    static String automationPrefix = "Automation."_s;
+    auto method = message->getString("method"_s);
+    if (!method.startsWith(automationPrefix)) {
+        WTFLogAlways("Unknown event domain: %s", method.utf8().data());
+        return;
+    }
+
+    auto eventName = method.substring(automationPrefix.length());
+    if (!m_globalEventSet.contains(eventName)) {
+        WTFLogAlways("Ignoring unsubscribed event %s", eventName.utf8().data());
+        return;
+    }
+
+
+    if (eventName == "logEntryAdded"_s) {
+        doLogEntryAdded(WTFMove(message));
+        return;
+    }
+}
+
+void Session::doLogEntryAdded(RefPtr<JSON::Object>&& message)
+{
+    // https://w3c.github.io/webdriver-bidi/#event-log-entryAdded
+    auto params = message->getObject("params"_s);
+    auto method = params->getString("method"_s);
+
+    String level;
+    if (method == "error"_s || method == "assert"_s)
+        level = "error"_s;
+    else if (method == "debug"_s || method == "trace"_s)
+        level = "debug"_s;
+    else if (method == "warn"_s)
+        level = "warn"_s;
+    else
+        level = "info"_s;
+
+    // WebDriver uses the ECMA time format, which is the number of milliseconds since the Unix epoch.
+    // https://tc39.es/ecma262/#sec-time-values-and-time-range
+    RefPtr<JSON::Value> timestampValue;
+    if (auto timestampOpt = params->getDouble("timestamp"_s))
+        timestampValue = JSON::Value::create(round(*timestampOpt));
+    else
+        timestampValue = JSON::Value::create(0);
+
+    StringBuilder textBuilder;
+
+    // TODO Support formatter string
+    // https://bugs.webkit.org/show_bug.cgi?id=282976
+    auto formattedArgs = params->getArray("args"_s);
+    if (!formattedArgs) {
+        formattedArgs = JSON::Array::create();
+        formattedArgs->pushString(params->getString("text"_s));
+    }
+
+    for (unsigned i = 0; i < formattedArgs->length(); ++i) {
+        auto arg = formattedArgs->get(i);
+        if (i)
+            textBuilder.append(' ');
+        if (arg->type() == JSON::Value::Type::String)
+            textBuilder.append(arg->asString());
+        else
+            textBuilder.append("null"_s);
```

Comment:
are we not supposed to serialize boolean or number values?

Reply:
Yes, there's a lengthy spec about serialization of the arguments. But after reviewing the `Automation.json` and the `WebAutomationSession` messages, they send only the final text string for a given console message.

So, for the sake of simplicity, we can remove the serialization steps _from this PR_, adding back in the follow-up PR's regarding multiple arguments/formatter strings (https://bugs.webkit.org/show_bug.cgi?id=282976) and actual serialization (https://bugs.webkit.org/show_bug.cgi?id=282977).

Maybe the serialization/formatting can be actually be implemented in the UIProcess instead of in the WebDriver service. @burg had some ideas in this direction for other BiDi methods/events too.